### ISSUE TITLE
[MASSEMBLY-851] "Unextractable" JAR files created by assembly:single

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-assembly-plugin#1049]
**[Tom Wieczorek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tom@bibbu.net)** opened **[MASSEMBLY-851](https://issues.apache.org/jira/browse/MASSEMBLY-851?redirect=false)** and commented

Under certain circumstances, `assembly:single` generates "unextractable" JAR files, because of some directory entries in the JAR file are not ending with a slash, hence they're treated as file entries.

If such a JAR file is tried to be extracted via the `jar` executable, an Exception is thrown and extraction is aborted:

```
java.io.FileNotFoundException: META-INF/maven (Is a directory)
        at java.io.FileOutputStream.open0(Native Method)
        at java.io.FileOutputStream.open(FileOutputStream.java:270)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
        at sun.tools.jar.Main.copy(Main.java:909)
        at sun.tools.jar.Main.extractFile(Main.java:1069)
        at sun.tools.jar.Main.extract(Main.java:981)
        at sun.tools.jar.Main.run(Main.java:311)
        at sun.tools.jar.Main.main(Main.java:1288)
```

A [sample project](https://github.com/twz123/broken-assembly) is on GitHub.

---

**Affects:** 3.0.0

**Attachments:**
- [mvn.log](https://issues.apache.org/jira/secure/attachment/12861449/mvn.log) (_3.41 kB_)

**Remote Links:**
- [Sample project
](https://github.com/twz123/broken-assembly)

0 votes, 5 watchers

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I haven't been able to reproduce it on my Windows machine.

> Under certain circumstances ...

Any clue which circumstances? Looks like it has happened a few times, not 100% reproducible.

**Comment by jira-importer** [comment]
**[Tom Wieczorek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tom@bibbu.net)** commented

Hmm, the Travis build also failed: https://travis-ci.org/twz123/broken-assembly/builds/216487972#L1220

Same result as on my local machine.

**Comment by jira-importer** [comment]
**[Tom Wieczorek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tom@bibbu.net)** commented

> Seems to be an OS issue...

Yeah, verified that by adding a [Windows build on AppVeyor](https://ci.appveyor.com/project/twz123/broken-assembly) that's working smoothly.

**Comment by jira-importer** [comment]
**[Tom Wieczorek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tom@bibbu.net)** commented

Okay, I enabled the three standard JDKs available on Travis:

OpenJDK 7:

```
java version "1.7.0_75"
OpenJDK Runtime Environment (IcedTea 2.5.4) (7u75-2.5.4-1~precise1)
OpenJDK 64-Bit Server VM (build 24.75-b04, mixed mode)
```

Oracle JDK 7:

```
java version "1.7.0_76"
Java(TM) SE Runtime Environment (build 1.7.0_76-b13)
Java HotSpot(TM) 64-Bit Server VM (build 24.76-b04, mixed mode
```

Oracle JDK 8:

```
java version "1.8.0_31"
Java(TM) SE Runtime Environment (build 1.8.0_31-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.31-b07, mixed mode)
```

All three builds fail: https://travis-ci.org/twz123/broken-assembly/builds/218227846

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Can reproduce:

```
[mosipov@bsd10 ~/Projekte/broken-assembly]$ JAVA_VERSION=1.7 mvn -V -q clean verify
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T17:41:47+01:00)
Maven home: /usr/local/apache-maven-3.3.9
Java version: 1.7.0_111, vendor: Oracle Corporation
Java home: /usr/local/openjdk7/jre
Default locale: de_DE, platform encoding: UTF-8
OS name: "freebsd", version: "10.3-release-p11", arch: "amd64", family: "unix"

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.github.twz123.brokenassembly.BrokenAssemblyIT
Executing [/usr/local/openjdk7/bin/jar, xf, /usr/home/mosipov/Projekte/broken-assembly/target/broken-assembly-0.0.1-SNAPSHOT-broken.jar]
java.io.FileNotFoundException: META-INF/maven (Is a directory)
        at java.io.FileOutputStream.open(Native Method)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:221)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:171)
        at sun.tools.jar.Main.copy(Main.java:853)
        at sun.tools.jar.Main.extractFile(Main.java:1043)
        at sun.tools.jar.Main.extract(Main.java:938)
        at sun.tools.jar.Main.run(Main.java:265)
        at sun.tools.jar.Main.main(Main.java:1262)
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.201 sec <<< FAILURE! - in com.github.twz123.brokenassembly.BrokenAssemblyIT
unjarBrokenAssembly(com.github.twz123.brokenassembly.BrokenAssemblyIT)  Time elapsed: 0.158 sec  <<< FAILURE!
java.lang.AssertionError:
Non-zero exit code from jar executable.
Expected: is <0>
     but: was <1>
        at com.github.twz123.brokenassembly.BrokenAssemblyIT.unjarBrokenAssembly(BrokenAssemblyIT.java:33)


Results :

Failed tests:
  BrokenAssemblyIT.unjarBrokenAssembly:33 Non-zero exit code from jar executable.
Expected: is <0>
     but: was <1>

Tests run: 1, Failures: 1, Errors: 0, Skipped: 0

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:2.19.1:verify (default) on project broken-assembly: There are test failures.
[ERROR]
[ERROR] Please refer to /usr/home/mosipov/Projekte/broken-assembly/target/failsafe-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

```
[mosipov@bsd10 ~/Projekte/broken-assembly]$ JAVA_VERSION=1.8 mvn -V -q clean verify
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T17:41:47+01:00)
Maven home: /usr/local/apache-maven-3.3.9
Java version: 1.8.0_112, vendor: Oracle Corporation
Java home: /usr/local/openjdk8/jre
Default locale: de_DE, platform encoding: UTF-8
OS name: "freebsd", version: "10.3-release-p11", arch: "amd64", family: "unix"

-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running com.github.twz123.brokenassembly.BrokenAssemblyIT
Executing [/usr/local/openjdk8/bin/jar, xf, /usr/home/mosipov/Projekte/broken-assembly/target/broken-assembly-0.0.1-SNAPSHOT-broken.jar]
java.io.FileNotFoundException: META-INF/maven (Is a directory)
        at java.io.FileOutputStream.open0(Native Method)
        at java.io.FileOutputStream.open(FileOutputStream.java:270)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
        at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
        at sun.tools.jar.Main.copy(Main.java:909)
        at sun.tools.jar.Main.extractFile(Main.java:1069)
        at sun.tools.jar.Main.extract(Main.java:981)
        at sun.tools.jar.Main.run(Main.java:311)
        at sun.tools.jar.Main.main(Main.java:1288)
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.28 sec <<< FAILURE! - in com.github.twz123.brokenassembly.BrokenAssemblyIT
unjarBrokenAssembly(com.github.twz123.brokenassembly.BrokenAssemblyIT)  Time elapsed: 0.21 sec  <<< FAILURE!
java.lang.AssertionError:
Non-zero exit code from jar executable.
Expected: is <0>
     but: was <1>
        at com.github.twz123.brokenassembly.BrokenAssemblyIT.unjarBrokenAssembly(BrokenAssemblyIT.java:33)


Results :

Failed tests:
  BrokenAssemblyIT.unjarBrokenAssembly:33 Non-zero exit code from jar executable.
Expected: is <0>
     but: was <1>

Tests run: 1, Failures: 1, Errors: 0, Skipped: 0

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:2.19.1:verify (default) on project broken-assembly: There are test failures.
[ERROR]
[ERROR] Please refer to /usr/home/mosipov/Projekte/broken-assembly/target/failsafe-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

Native unzip:

```
[mosipov@bsd10 ~/Projekte/broken-assembly]$ unzip -t target/broken-assembly-0.0.1-SNAPSHOT.jar
Archive:  target/broken-assembly-0.0.1-SNAPSHOT.jar
    testing: META-INF/MANIFEST.MF        OK
    testing: com/github/twz123/brokenassembly/Foo.class  OK
    testing: META-INF/maven/     OK
    testing: META-INF/maven/com.github.twz123/   OK
    testing: META-INF/maven/com.github.twz123/broken-assembly/   OK
    testing: META-INF/maven/com.github.twz123/broken-assembly/pom.xml    OK
    testing: META-INF/maven/com.github.twz123/broken-assembly/pom.properties     OK
No errors detected in compressed data of target/broken-assembly-0.0.1-SNAPSHOT.jar.
[mosipov@bsd10 ~/Projekte/broken-assembly]$ unzip -t target/broken-assembly-0.0.1-SNAPSHOT-broken.jar
Archive:  target/broken-assembly-0.0.1-SNAPSHOT-broken.jar
    testing: META-INF/MANIFEST.MF        OK
    testing: com/github/twz123/brokenassembly/Foo.class  OK
    testing: META-INF/maven/com.github.twz123/broken-assembly/pom.xml    OK
    testing: META-INF/maven/com.github.twz123/broken-assembly/pom.properties     OK
    testing: META-INF/maven      OK
    testing: META-INF/maven/com.github.twz123/broken-assembly    OK
    testing: META-INF/maven/com.github.twz123    OK
No errors detected in compressed data of target/broken-assembly-0.0.1-SNAPSHOT-broken.jar.
[mosipov@bsd10 ~/Projekte/broken-assembly]$ cd target/
[mosipov@bsd10 ~/Projekte/broken-assembly/target]$ unzip broken-assembly-0.0.1-SNAPSHOT.jar
Archive:  broken-assembly-0.0.1-SNAPSHOT.jar
d META-INF
 extracting: META-INF/MANIFEST.MF
d com
d com/github
d com/github/twz123
d com/github/twz123/brokenassembly
 extracting: com/github/twz123/brokenassembly/Foo.class
unzip: skipping non-regular entry 'META-INF/maven'
unzip: skipping non-regular entry 'META-INF/maven/com.github.twz123'
unzip: skipping non-regular entry 'META-INF/maven/com.github.twz123/broken-assembly'
 extracting: META-INF/maven/com.github.twz123/broken-assembly/pom.xml
 extracting: META-INF/maven/com.github.twz123/broken-assembly/pom.properties
[mosipov@bsd10 ~/Projekte/broken-assembly/target]$ unzip  broken-assembly-0.0.1-SNAPSHOT-broken.jar
Archive:  broken-assembly-0.0.1-SNAPSHOT-broken.jar
 extracting: META-INF/MANIFEST.MF
d META-INF
d com
d com/github
d com/github/twz123
d com/github/twz123/brokenassembly
 extracting: com/github/twz123/brokenassembly/Foo.class
 extracting: META-INF/maven/com.github.twz123/broken-assembly/pom.xml
 extracting: META-INF/maven/com.github.twz123/broken-assembly/pom.properties
unzip: skipping non-regular entry 'META-INF/maven'
unzip: skipping non-regular entry 'META-INF/maven/com.github.twz123/broken-assembly'
unzip: skipping non-regular entry 'META-INF/maven/com.github.twz123'
```

**Comment by jira-importer** [comment]
**[Tom Wieczorek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tom@bibbu.net)** commented

> unzip: skipping non-regular entry 'META-INF/maven'

That's what I tried to describe in the initial bug report. There are some malformed directory entries in the JAR. I suppose the reason why this bug won't show up on Windows is due to the different behavior of the filesystem.

**Comment by jira-importer** [comment]
**[Tom Wieczorek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tom@bibbu.net)** commented

Okay, spent some time debugging: The behavior is triggered when the Assembly Plugin directly repackages JAR files into the assembly JAR.

Apache Commons Compress is used to read the source JAR file. When reading the external attributes of the directory entries from the source JAR file, for some reason, the symlink flag is `true` for the entries below `META-INF/`. That triggers, of course, special treatment in the Plexus Archiver, which removes the trailing slash from the direcory entry names in the destination JAR file. This, in turn, triggers the error in the jar binary, because it treats the trailing slash as indicator if a ZIP file entry is a file or a directory.

This explains why this issue doesn't show up on Windows, since the Windows filesystems don't support symlinks and the code paths in question aren't used.

So the question is: Is it a bug inside Commons Compress when reading the external attributes or is it already wrong inside the source JAR file, which means that I'd need to dig through the Maven Jar Plugin, probably landing again inside Plexus Archiver and Commons Compress...

This is getting somewhat tricky...

**Comment by jira-importer** [comment]
**[Tom Wieczorek](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=tom@bibbu.net)** commented

Not an issue with the Assembly Plugin, so this issue can be closed. The Assembly plugin just consumes a slightly malformed JAR produced by the Maven JAR Plugin. I just hit this problem because I didn't bother to upgrade the default versions of the stock Maven plugins. Maven 3.3.9 uses by default the rather ancient version 2.4 of the Maven JAR Plugin. This version in turn uses Plexus Archiver 2.1. That version has a [bug that has already been fixed in 2012](https://github.com/codehaus-plexus/plexus-archiver/commit/a159f8355f957966e011b523a1698ee17900c7fc) in the next patch version 2.1.1.

Hence, the solution is so easy: Use Maven JAR Plugin in a version >= 2.5. Version 2.5 ships with Plexus Archiver 2.4.4 ([MJAR-168](https://issues.apache.org/jira/browse/MJAR-168)).

I added two profiles to the sample project, just for reference. The first uses JAR plugin 2.5, the second uses JAR plugin 2.4 along with Plexus Archiver 2.1.1. Both profiles pass the tests on my machine.
